File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashMap<Character,Integer>mp= new HashMap<>();

        int left=0,maxLen=0;
        for(int r=0;r<s.length();r++){
            char curr=s.charAt(r);

            if(mp.containsKey(curr)&& left<=mp.get(curr)){
                left=mp.get(curr)+1;
            }
            mp.put(curr,r);
            maxLen=Math.max(maxLen,r-left+1);
        }
    return maxLen;
    }
}

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        vector<int> arr(256,-1);
        int l=0,res=0;
        for(int r=0;r<s.size();r++){
            int c=s[r];
            if(arr[c]>=l){
                l=arr[c]+1;
            }
            arr[c]=r;
            res=max(res,r-l+1);
        }
    return res;
    }
};